### PR TITLE
Entity, DAO and Service for Faceted Search Classification

### PR DESCRIPTION
**Pull request by PenghaiZhang**
#764 
##### Checklist

- [x] the [contributor license agreement] is signed
- [x] commit message follows [commit guidelines]

##### Description of change

The initial work creates a new entity which maps the `faceted_search_classification` table, and creates the DAO and Service layers for Faceted Search Classification.

### CODE REVIEW

File: Source/Plugins/Core/com.equella.core/src/com/tle/core/facetedsearch/bean/FacetedSearchClassification.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+  @Column(nullable = false)
+  private String name;
+
+  @Column(nullable = false)
+  private String schemaNode;
+
+  private int maxResults;
+
+  private int orderIndex;
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/src/com/tle/core/facetedsearch/bean/FacetedSearchClassification.java:34]: I am not sure if `maxResults` and `orderIndex` can be null or not so at the moment I assume they could be null.

File: Source/Plugins/Core/com.equella.core/src/com/tle/core/facetedsearch/dao/FacetedSearchClassificationDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    return getHibernateTemplate()
+        .find(
+            "from FacetedSearchClassification where institution_id = ?",
+            CurrentInstitution.get().getUniqueId());
+  }
+
+  @Transactional
+  @Override
+  public void deleteClassification(FacetedSearchClassification facetedSearchClassification) {
+    delete(facetedSearchClassification);
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/src/com/tle/core/facetedsearch/dao/FacetedSearchClassificationDaoImpl.java:54]: There are two ways of doing a delete in oEQ. One is calling the `delete` implemented in `GenericDaoImpl` with an entity as parameter, and the other one is using Hibernate Query. Both are used in quite a few places.

File: Source/Plugins/Core/com.equella.core/src/com/tle/core/facetedsearch/bean/FacetedSearchClassification.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  @Id
+  @GeneratedValue(strategy = GenerationType.AUTO)
+  private long id;
+
+  @Column(nullable = false)
+  private long institutionId;
+
+  @Column @JsonIgnore private Date dateCreated;
+
+  @Column @JsonIgnore private Date dateModified;
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/src/com/tle/core/facetedsearch/bean/FacetedSearchClassification.java:24]: I think there is no need to return `dateCreated` and `dateModified` to front-end so I add `@JsonIgnore` here.

**Comment by PenghaiZhang** [comment]
Screenshot from 2020-04-23 11-08-46
Screenshot from 2020-04-23 11-47-16

**Comment by PenghaiZhang** [comment]
Export
Screenshot from 2020-04-23 11-08-46

Import
Screenshot from 2020-04-23 11-47-16

Data Migration
Screenshot from 2020-04-23 12-15-38
Screenshot from 2020-04-23 12-16-02

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/FacetedSearch/FacetedSearchClassificationResource.scala [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  @GET
+  @Path("facetedsearch/classification/{id}")
+  @ApiOperation(
+    value = "Retrieve one faceted search classification",
+    notes = "This endpoint is used to retrieve one faceted search classification.",
+    response = classOf[FacetedSearchClassification]
+  )
+  def get(@ApiParam(value = "ID") @PathParam("id") id: Long): Response = {
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/FacetedSearch/FacetedSearchClassificationResource.scala:65]: Isn't `value` redundant?

```suggestion
  def get(@ApiParam("ID") @PathParam("id") id: Long): Response = {
```

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/FacetedSearch/FacetedSearchClassificationResource.scala [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    notes = "This endpoint is used to add one faceted search classification.",
+    response = classOf[FacetedSearchClassification]
+  )
+  def add(classification: FacetedSearchClassification): Response = {
+    privilegeChecker.checkAuthorised()
+    validate(classification) match {
+      case Left(errors) => ApiErrorResponse.badRequest(errors: _*)
+      case Right(_) =>
+        classification.setDateCreated(new Date())
+        classification.setDateModified(new Date())
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/FacetedSearch/FacetedSearchClassificationResource.scala:86]: Shame the DAO is not doing this, or that it's not being done with hibernate hooks/events. However if oEQ doesn't already do that, then fine....

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/FacetedSearch/FacetedSearchClassificationResource.scala [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  @PUT
+  @Path("facetedsearch/classification/{id}")
+  @ApiOperation(
+    value = "Update one faceted search classification",
+    notes = "This endpoint is used to update one faceted search classification.",
+    response = classOf[FacetedSearchClassification]
+  )
+  def update(@ApiParam(value = "ID") @PathParam("id") id: Long,
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/FacetedSearch/FacetedSearchClassificationResource.scala:100]:
```suggestion
  def update(@ApiParam("ID") @PathParam("id") id: Long,
```

File: autotest/OldTests/src/test/java/com/tle/webtests/test/webservices/rest/FacetedSearchClassificationApiTest.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  @Override
+  protected void addOAuthClients(List<Pair<String, String>> clients) {
+    clients.add(new Pair<String, String>(OAUTH_CLIENT_ID, "AutoTest"));
+  }
+
+  @BeforeClass
+  @Override
+  public void registerClients() throws Exception {
+    super.registerClients();
+    JsonNode node = mapper.readTree(Attachments.get("facetedsearchclassification.json"));
```
Comment (edalex-ian) [inline autotest/OldTests/src/test/java/com/tle/webtests/test/webservices/rest/FacetedSearchClassificationApiTest.java:34]: Sees odd to need to use something called `Attachments` for this. Normally you'd just use the `classloader.getResourceAsStream(<filename>)`. But maybe this is the oEQ way?

**Pull request merged**